src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,926 @@
+-- 'isEmpty' . connects == 'all' 'isEmpty'
+-- @
+connects :: [Graph a] -> Graph a
+connects = fromMaybe empty . foldr1Safe connect
+{-# INLINE [1] connects #-}
+
+-- | Generalised 'Graph' folding: recursively collapse an 'Graph' by applying
+-- the provided functions to the leaves and internal nodes of the expression.
+-- The order of arguments is: empty, vertex, overlay and connect.
+-- Complexity: /O(s)/ applications of given functions. As an example, the
+-- complexity of 'size' is /O(s)/, since all functions have cost /O(1)/.
+--
+-- @
+-- foldg 'empty' 'vertex'        'overlay' 'connect'        == id
+-- foldg 'empty' 'vertex'        'overlay' ('flip' 'connect') == id
+-- foldg 1     ('const' 1)     (+)     (+)            == 'size'
+-- foldg True  ('const' False) (&&)    (&&)           == 'isEmpty'
+-- foldg False (== x)        (||)    (||)           == 'hasVertex' x
+-- @
+foldg :: b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> Graph a -> b
+foldg e v o c (UG g) = go g
+  where
+    go G.Empty         = e
+    go (G.Vertex  x  ) = v x
+    go (G.Overlay x y) = o (go x) (go y)
+    go (G.Connect x y) = c (go x) (go y)
+{-# INLINE [0] foldg #-}
+
+-- TODO: Add rule (?) 
+    -- foldg e v o c (Connect x y) = c (foldg e v o c y) (foldg e v o c x)
+{-# RULES
+"foldg/Empty"   forall e v o c.
+    foldg e v o c (UG G.Empty) = e
+"foldg/Vertex"  forall e v o c x.
+    foldg e v o c (UG (G.Vertex x)) = v x
+"foldg/Overlay" forall e v o c x y.
+    foldg e v o c (UG (G.Overlay x y)) = o (foldg e v o c (UG x)) (foldg e v o c (UG y))
+"foldg/Connect" forall e v o c x y.
+    foldg e v o c (UG (G.Connect x y)) = c (foldg e v o c (UG x)) (foldg e v o c (UG y))
+
+"foldg/overlays" forall e v o c xs.
+    foldg e v o c (overlays xs) = fromMaybe e (foldr (maybeF o . foldg e v o c) Nothing xs)
+"foldg/connects" forall e v o c xs.
+    foldg e v o c (connects xs) = fromMaybe e (foldr (maybeF c . foldg e v o c) Nothing xs)
+ #-}
+
+-- | The 'isSubgraphOf' function takes two graphs and returns 'True' if the
+-- first graph is a /subgraph/ of the second.
+-- Complexity: /O(s + m * log(m))/ time. Note that the number of edges /m/ of a
+-- graph can be quadratic with respect to the expression size /s/.
+--
+-- @
+-- isSubgraphOf 'empty'         x             ==  True
+-- isSubgraphOf ('vertex' x)    'empty'         ==  False
+-- isSubgraphOf x             ('overlay' x y) ==  True
+-- isSubgraphOf ('overlay' x y) ('connect' x y) ==  True
+-- isSubgraphOf ('path' xs)     ('circuit' xs)  ==  True
+-- isSubgraphOf ('edge' x y)    ('edge' y x)    ==  True
+-- isSubgraphOf x y                         ==> x <= y
+-- @
+isSubgraphOf :: Ord a => Graph a -> Graph a -> Bool
+isSubgraphOf x y = SR.isSubgraphOf (toSymmetricRelation x) (toSymmetricRelation y)
+{-# NOINLINE [1] isSubgraphOf #-}
+{-# RULES "isSubgraphOf/Int" isSubgraphOf = isSubgraphOfIntR #-}
+
+-- Like 'isSubgraphOf' but specialised for graphs with vertices of type 'Int'.
```

Comment:
This isn't currently specialised, so this comment is wrong. Please add a TODO to fix this.

Reply:
Yes, you're right! I forgot this ones too!